File: src/main/java/com/dorpeled/pterodactylcommandsmod/util/PterodactylUrlBuilder.java
package com.dorpeled.pterodactylcommandsmod.util;

import com.dorpeled.pterodactylcommandsmod.config.PterodactylCommandsConfig;

import java.util.ArrayList;
import java.util.List;

public class PterodactylUrlBuilder {
    private static PterodactylUrlBuilder instance;
    private String baseUrl;
    private String endpoint;
    private List<String> params;

    private PterodactylUrlBuilder() {
        this.baseUrl = PterodactylCommandsConfig.BASE_URL.get();
        params = new ArrayList<>();
    }

    public static PterodactylUrlBuilder getInstance() {
        if (instance == null) {
            instance = new PterodactylUrlBuilder();
        }
        return instance;
    }

    public PterodactylUrlBuilder endpoint(String endpoint) {
        this.endpoint = endpoint;
        return this;
    }

    public PterodactylUrlBuilder    param(String param) {
        this.params.add(param);
        return this;
    }

    public String build() {
        StringBuilder urlBuilder = new StringBuilder(baseUrl);
        urlBuilder.append(endpoint);

        for (String param : params) {
            urlBuilder.append("/");
            urlBuilder.append(param);
        }

        params.clear();
        endpoint = null;

        return urlBuilder.toString();
    }

    public String getServerUrl() {
        return this.endpoint("/api/client/servers")
                .param(PterodactylCommandsConfig.SERVER_ID.get())
                .build();
    }

    public String getScheduleListUrl() {
        return this.endpoint("/api/client/servers")
                .param(PterodactylCommandsConfig.SERVER_ID.get())
                .param("schedules")
                .build();
    }

    public String getScheduleExecuteUrl(String scheduleId) {
        return this.endpoint("/api/client/servers")
                .param(PterodactylCommandsConfig.SERVER_ID.get())
                .param("schedules")
                .param(scheduleId)
                .param("execute")
                .build();
    }
}

File: src/main/kotlin/com/dorpeled/pterodactylcommandsmod/util/PterodactylUrlBuilder.kt
package com.dorpeled.pterodactylcommandsmod.util

import com.dorpeled.pterodactylcommandsmod.config.PterodactylCommandsConfig

class PterodactylUrlBuilder private constructor() {
    private val baseUrl: String = PterodactylCommandsConfig.BASE_URL.get() ?: throw IllegalStateException("Base URL not configured")
    private var endpoint: String = "/api/client/servers"
    private val params: MutableList<String> = mutableListOf()

    private fun param(param: String): PterodactylUrlBuilder = apply { params.add(param) }

    private fun build(): String = StringBuilder(baseUrl).apply {
        append(endpoint)
        params.forEach { append("/").append(it) }
        params.clear()
    }.toString()

    fun getServerUrl(): String = param(PterodactylCommandsConfig.SERVER_ID.get()).build()

    fun getScheduleListUrl(): String = param(PterodactylCommandsConfig.SERVER_ID.get()).param("schedules").build()

    fun getScheduleExecuteUrl(scheduleId: String): String = param(PterodactylCommandsConfig.SERVER_ID.get()).param("schedules").param(scheduleId).param("execute").build()

    companion object {
        val instance: PterodactylUrlBuilder by lazy { PterodactylUrlBuilder() }
    }
}
